ta/pkcs11/src/processing.c

```diff
@@ -747,3 +784,185 @@ enum pkcs11_rc entry_release_active_processing(struct pkcs11_client *client,
+	struct pkcs11_object *wrapping_key = NULL;
+	struct pkcs11_object *key = NULL;
+	struct obj_attrs *req_attrs = NULL;
+	uint32_t wrapping_key_handle = 0;
+	uint32_t key_handle = 0;
+	uint32_t size = 0;
+	void *key_data = NULL;
+	size_t key_sz = 0;
+	void *out_buf = params[2].memref.buffer;
+	uint32_t out_size = params[2].memref.size;
+	enum processing_func function = PKCS11_FUNCTION_WRAP;
+
+	if (!client || ptypes != exp_pt ||
+	    (out_size && !out_buf))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &wrapping_key_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_get(&ctrlargs, &key_handle, sizeof(uint32_t));
+	if (rc)
+		return rc;
+
+	rc = serialargs_alloc_get_one_attribute(&ctrlargs, &proc_params);
+	if (rc)
+		goto out;
+
+	if (serialargs_remaining_bytes(&ctrlargs)) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	/* Return error if processing already active */
+	rc = get_ready_session(session);
+	if (rc)
+		goto out;
+
+	/* Check wrapping key handle */
+	wrapping_key = pkcs11_handle2object(wrapping_key_handle, session);
+	if (!wrapping_key) {
+		rc = PKCS11_CKR_WRAPPING_KEY_HANDLE_INVALID;
+		goto out;
+	}
+
+	/* Check key handle */
+	key = pkcs11_handle2object(key_handle, session);
+	if (!key) {
+		rc = PKCS11_CKR_KEY_HANDLE_INVALID;
+		goto out;
+	}
+
+	/*
+	 * The wrapping key and key to be wrapped shouldn't be same.
+	 * PKCS#11 spec doesn't explicitly state that but logically this isn't
+	 * a use case and also acts as an attack vector, so explicitly
+	 * disallow this.
+	 */
+	if (key == wrapping_key) {
+		rc = PKCS11_CKR_ARGUMENTS_BAD;
+		goto out;
+	}
+
+	/* Check if mechanism can be used for wrapping function */
+	rc = check_mechanism_against_processing(session, proc_params->id,
+						function,
+						PKCS11_FUNC_STEP_INIT);
+	if (rc)
+		goto out;
+
+	/* Set the processing state to active */
+	rc = set_processing_state(session, function, wrapping_key, NULL);
+	if (rc)
+		goto out;
+
+	/*
+	 * Check if wrapping key has CKA_WRAP set and its key type is
+	 * compatible with the mechanism passed
+	 */
+	rc = check_parent_attrs_against_processing(proc_params->id, function,
+						   wrapping_key->attributes);
+	if (rc) {
+		/*
+		 * CKR_KEY_FUNCTION_NOT_PERMITTED is not in the list of errors
+		 * specified with C_Wrap() in the specification. So
+		 * return the next most appropriate error.
+		 */
+		if (rc == PKCS11_CKR_KEY_FUNCTION_NOT_PERMITTED)
+			rc = PKCS11_CKR_WRAPPING_KEY_TYPE_INCONSISTENT;
+
+		goto out;
+	}
+
+	/* Check access of wrapping key */
+	rc = check_access_attrs_against_token(session,
+					      wrapping_key->attributes);
+	if (rc)
+		goto out;
+
+	switch (get_class(key->attributes)) {
+	case PKCS11_CKO_SECRET_KEY:
+	case PKCS11_CKO_PRIVATE_KEY:
```

Comment:
private keys are not supported with current code.

Reply:
will remove